Implement simpler data fetch from OCLC Classify service

mwbenowitz opened the pull request:
After several discussions and a review of the overall architecture, it seems more prudent to simply have this service fetch data and return it for further processing, rather than attempt to manage data merging and updates. As this is an intermediate step that risked losing sight of what changes were made to the data as well as loosing the ability to trace the source of all data in the system.

This update makes the service simpler, enabling queries to be made against the Classify service using either a title/author search or a search by identifier (oclc#, lccn, etc.) The function transforms the received XML document into an object that complies with the SFR data model (as defined in the dataModel.py file in this repository). That object is then returned to a stream for persistence to the db.

mwbenowitz committed:
SFR-206 Simplify OCLC Classify service function

This update removes the merging and data handling aspects of the
OCLC Classify service. It changes this function to simply
retrieve Classify records from OCLC, format them to meet the SFR
data model and return them to the data stream.

This simplifies this process and removes issues regarding how and
when data is being merged at this phase

mwbenowitz committed:
Fix tests after removing methods

mwbenowitz committed:
Improved error handling with input parameters

mwbenowitz committed:
Add comments + method/class descriptions

mwbenowitz committed:
Make Kinesis writer more generic

Remove specific method/status from kinesis output and allow that
to be set by the calling method

seanredmond commented on service.py:
```diff
@@ -24,9 +25,6 @@ def handler(event, context):
-    # Method to be invoked goes here
-    # TODO Implement oauth checking
-
     results = parseRecords(records)
```
Since all parseRecords() does is `list(map(parseRecord, records))`, it seems you could skip the extra function call.

mwbenowitz replied:
You're right, but I think that having the one-liner makes reading this slightly easier (for me at least), I'm open to disagreement on that though

leonardr replied:
FWIW another reason to do this is so you can mock `parseRecords` separately during tests.

seanredmond commented on service.py:
```diff
@@ -36,11 +34,15 @@ def handler(event, context):
+    """Simple method to parse list of records and process each entry."""
     logger.debug("Parsing Messages")
     results = list(map(parseRecord, records))
     return results
```
Just a style thing, but I'm a fan of fewer variables, i.e. `return list(map(parseRecord, records))` (but see comment above about maybe getting rid of this function.

Also, I would have expected a list comprehension like ` return [parseRecord(r) for r in records]`. There's not much of a _technical_ difference -- but when in Rome...

mwbenowitz replied:
No, but I switched it to list comprehension. I'm not sure why I started using `map()` but I was experimenting with it in an earlier version of the boilerplate code this is based on and its a carry-over here.

seanredmond commented on service.py:
```diff
@@ -52,14 +54,5 @@ def parseRecord(encodedRec):
-        logger.info('This record is not for this stage, return for further processing')
-        return False
-
     result = enhanceRecord(record)
```
I think I would put this and the return in the `try`, take `return False` out of the `excepts` and make _it_ the last-ditch return

Also `enhanceRecord()` does a lot of error checking and returning `False`. It might read more cleanly if `enhanceRecord()` just threw errors which were caught here, in one place, and logged.

mwbenowitz replied:
Both good ideas I think for making the control flow simpler and the methods easier to read.

seanredmond commented on lib/enhancer.py:
```diff
@@ -1,80 +1,68 @@
-        KinesisOutput.putRecord(mergedData)
+        classifyData = classifyRecord(searchType, searchFields, workUUID)
+
+        if classifyData is None:
```
Probably beyond the scope is this PR, but this feels like a potentially confusion side-effect, as does the `putRecord()` As I understand the return values `False` means there was some kind of error, `True` means one kind of record was stored, and `None` means either nothing happened  or multiple records were queued for later? Maybe this needs to be broken down another way?

mwbenowitz replied:
Agreed. I'm going to refactor this so instead of returning `False` or `None` it raises an exception in those cases, which is caught and logged in a similar fashion to what you suggested above

mwbenowitz committed:
Improve/simplify control flow

Implements several updates to raise/catch custom errors to
simplify flow of function. This now will simply process the record
or raise a relevant error, instead of returning None/False.

mwbenowitz committed:
Remove extraneous return

leonardr reviewed (approved):
I'm sorry it took me so long to review this branch. There's a big improvement here in overall strategy.
